### PR TITLE
Added support for . (any character) token in grammar engine.

### PR DESCRIPTION
**Pull request by HanClinto**
Low priority feature. Consider this more of a suggestion than a request. :)

As came up in discussion of #6441, I wanted a way to create a grammar that ensured a minimum response length. It seemed prudent to add support for a "." character that would match on any generated token -- without it, I used the string `[^\x00]`, which matches on any non-null character (which is very nearly the same thing).

I don't know if this character be useful in too many other situations or not, so feel free to leave this one out if you don't think it's worthy.

I didn't add this token to the grammar tests, because frankly I haven't really been able to wrap my head around them. I would still like to eventually get around to writing some end-to-end / integration tests for the grammar engine that are a bit easier to grok and extend, but unless otherwise requested, I'll leave that exercise for another PR.

Example usage:
```
./main -m ./models/llama-2-7b.Q4_0.gguf -e -r "\n" --grammar "root ::= [^\n].+" -p "My favorite flavor is "
```

Example output:
```
 My favorite flavor is 🍌. surely you know that I love to eat meat, I'm a real carnivore, I like to eat pork, chicken and beef. I can not imagine my life without meat, I also like to eat seafood.
 ```

### CODE REVIEW

**Comment by ggerganov** [comment]
> I didn't add this token to the grammar tests, because frankly I haven't really been able to wrap my head around them. I would still like to eventually get around to writing some end-to-end / integration tests for the grammar engine that are a bit easier to grok and extend, but unless otherwise requested, I'll leave that exercise for another PR.

Matching any character seems it could be a useful addition, but I agree it would be better to first focus on improving grammar tests (and potentially performance). We can revisit this addition at a bit later point

**Comment by HanClinto** [comment]
> but I agree it would be better to first focus on improving grammar tests (and potentially performance). We can revisit this addition at a bit later point

Sounds great, thanks! That's where I'll turn my attention then -- I've done some profiling, and have some ideas in the works for how to improve the grammar parser. Next step will be to add profiling to the grammar engine (I ned to investigate the current state of benchmarks that include grammars). Meanwhile, this addition can stay here and we can return to it whenever we feel like it.

Thank you!

**Comment by ggerganov** [comment]
Awesome! The grammar functionality is a great feature and would be nice to get some extra attention. I sent you a collaborator invite, if you feel like helping out (no pressure if you don't have time / resources, this is mainly a token of appreciation at this point)

**Comment by HanClinto** [comment]
> Awesome! The grammar functionality is a great feature and would be nice to get some extra attention. I sent you a collaborator invite, if you feel like helping out (no pressure if you don't have time / resources, this is mainly a token of appreciation at this point)

Wow, I am honored -- thank you very much!! I will do my best to not abuse the privilege.

I agree about the grammar functionality being very powerful. I think that verifiable correctness is going to be one of the big ways that local LLMs can really gain some usefulness -- I started doing some experiments a couple of months ago with text-to-SQL generation and I think that using grammars to ensure syntactically-correct SQL queries (even tuned for a person's specific database schema) offer a lot of potential to expand usefulness of LLMs. That's what started me down this whole road of digging into grammars on llama.cpp, and I'm excited to see what the future holds for it.

Thanks for everything!

**Review by ggerganov — APPROVED** [review]
@HanClinto Feel free to merge this if it is ready

**Comment by HanClinto** [comment]
> @HanClinto Feel free to merge this if it is ready

Thank you! I'm much more familiar with the grammar engine now than I was when I first wrote this, so I'd like to try to look it all over again with fresh eyes. 

Overall I'm feeling much better about making changes like this to the grammar engine now that we have integration test coverage.

@ochafik not sure what your availability is these days, but wouldn't mind your critique at some point as well.

File: common/grammar-parser.cpp
Comment (HanClinto) [inline common/grammar-parser.cpp:193]: Currently wondering if it's worth adding a whole new LLAMA_GRETYPE for "." just for this.

The alternative is that when we're parsing ".", instead of inserting `LLAMA_GRETYPE_CHAR_ANY` into `out_elements`, we could instead add TWO elements -- a `LLAMA_GRETYPE_CHAR` set to 0, followed immediately by a `LLAMA_GRETYPE_CHAR_RNG_UPPER` set to max. This should have the same functionality as `LLAMA_GRETYPE_CHAR_ANY`, but at the cost of having extra cruft in the data structure.

This would effectively change every "." into "[\x0000-\xffff]" (and be handled identically).

That feels like the RISC method, and the current implementation is the CISC method. I'm not sure which is better, but just reviewing and thinking out loud here. Now that I type it all out, I kinda' prefer the cleanliness of the `LLAMA_GRETYPE_CHAR_ANY` method -- even if we have to add a new entry to the enum.

**Comment by HanClinto** [comment]
TODO before merge:

* [ ] Add "." symbol to integration tests

File: common/grammar-parser.cpp
Comment (ochafik) [reply common/grammar-parser.cpp:193]: I would lean towards keeping things as simple as possible (just one more of many parser-rewritten artefacts), esp. since I doubt there's any measurable runtime difference between the two approaches.

Comment (ochafik) [reply common/grammar-parser.cpp:193]: That said this doesn't add much complexity so also happy to see this merged as it is ;-)

**Comment by mofosyne** [comment]
Noting that it appears there is a general agreement to merge this PR, but just waiting on someone to add "." symbol to integration tests.

**Comment by HanClinto** [comment]
Rebased on `master`, integration tests added in 9e305134e22cf9edbb2420730a7792a3f53e7a2f -- ready for final review and merge!

File: tests/test-grammar-integration.cpp
Comment (ochafik) [inline tests/test-grammar-integration.cpp:221]: 🤗🤪

**Review by ochafik — APPROVED** [review]
Looks good!

**Pull request merged**
